packages/backend/src/inMemoryAnalytics.ts

```diff
@@ -0,0 +1,76 @@
+import { tableFromJSON, tableToIPC } from 'apache-arrow';
+import { Database } from 'duckdb-async';
+import Logger from './logging/logger';
+import { wrapOtelSpan } from './utils';
+
+type InMemoryJSONRow = Record<string, unknown>;
+type InMemoryJSONTableMap = Record<string, InMemoryJSONRow[]>;
+
+export interface InMemoryDatabaseProvisioningOptions {
+    tables?: InMemoryJSONTableMap;
+}
+
+/**
+ * Creates, and optionally provisions a new in-memory DuckDB instance.
+ *
+ * If table information is provided for provisioning, we use apache-arrow to load
+ * said table information into DuckDB, in the format { tableName: rows[] }
+ *
+ * We use a thin wrapper around DuckDB that exposes its methods as promises.
+ * See: https://www.npmjs.com/package/duckdb-async
+ */
+async function createDuckDbDatabase({
+    tables,
+}: InMemoryDatabaseProvisioningOptions) {
+    const db = await Database.create(':memory:');
+    await db.exec(`
+        INSTALL arrow;
+        LOAD arrow;
+
+        SET threads TO 1;
+        SET memory_limit = '512MB';
```

Comment:
Arbitrary choice on my part; might require some testing to reach a reasonable limit.